tests/src/junit/java/net/neoforged/neoforge/unittest/transfer/ItemTests.java

```diff
@@ -0,0 +1,261 @@
+        if (stack != container.getItem(0)) throw new AssertionError("Stack should have stayed the same.");
+        if (!stackEquals(stack, newResource, 5)) throw new AssertionError("Failed to update stack components or count.");
+    }
+
+    @Test
+    public void testContainerWrappers() {
+        ItemResource emptyBucket = ItemResource.of(Items.BUCKET);
+        TestWorldlyContainer testContainer = new TestWorldlyContainer();
+        checkComparatorOutput(testContainer);
+
+        // Create a few wrappers.
+        var unsidedWrapper = VanillaContainerWrapper.of(testContainer);
+        var downWrapper = new WorldlyContainerWrapper(testContainer, Direction.DOWN);
+        var upWrapper = new WorldlyContainerWrapper(testContainer, Direction.UP);
+
+        // Make sure querying a new wrapper returns the same one.
+        if (VanillaContainerWrapper.of(testContainer) != unsidedWrapper) throw new AssertionError("Wrappers should be ==.");
+
+        for (int iter = 0; iter < 2; ++iter) {
+            // First time, abort.
+            // Second time, commit.
+            try (Transaction transaction = Transaction.open(null)) {
+                // Insert bucket from down - should fail.
+                if (downWrapper.insert(emptyBucket, 1, transaction) != 0) throw new AssertionError("Bucket should not have been inserted.");
+                // Insert bucket unsided - should go in slot 1 (canPlaceItem returns false for slot 0).
+                if (unsidedWrapper.insert(emptyBucket, 1, transaction) != 1) throw new AssertionError("Failed to insert bucket.");
+                if (!testContainer.getItem(0).isEmpty()) throw new AssertionError("Slot 0 should have been empty.");
+                if (!stackEquals(testContainer.getItem(1), Items.BUCKET, 1)) throw new AssertionError("Slot 1 should have been a bucket.");
+                // The bucket should be extractable from any side but the top.
+                if (!emptyBucket.equals(ResourceHandlerUtil.findExtractableResource(unsidedWrapper, r -> true, transaction))) throw new AssertionError("Bucket should be extractable from unsided wrapper.");
+                if (!emptyBucket.equals(ResourceHandlerUtil.findExtractableResource(downWrapper, r -> true, transaction))) throw new AssertionError("Bucket should be extractable from down wrapper.");
+                if (ResourceHandlerUtil.findExtractableResource(upWrapper, r -> true, transaction) != null) throw new AssertionError("Bucket should NOT be extractable from up wrapper.");
+
+                if (iter == 1) {
+                    // Commit the second time only.
+                    transaction.commit();
+                }
+            }
+        }
+
+        // Check commit.
+        if (!testContainer.getItem(0).isEmpty()) throw new AssertionError("Slot 0 should have been empty.");
+        if (!testContainer.getItem(1).is(Items.BUCKET) || testContainer.getItem(1).getCount() != 1) throw new AssertionError("Slot 1 should have been a bucket.");
```

Comment:
I think this would be a better message as: "Slot 1 should have been a single bucket."